sync-core/src/main/java/com/cloudant/sync/datastore/BasicDatastore.java

```diff
@@ -791,53 +805,69 @@ private BasicDocumentRevision deleteDocumentInQueue(SQLDatabase db, final String
-                            // Previous revision to be deleted could be winner revision ("current" == true),
-                            // or a non-winner leaf revision ("current" == false), the new inserted
-                            // revision must have the same flag as it previous revision.
-                            // Deletion of non-winner leaf revision is mainly used when resolving
-                            // conflicts.
-                            InsertRevisionOptions options = new InsertRevisionOptions();
-                            options.docNumericId = preRevision.getInternalNumericId();
-                            options.revId = newRevisionId;
-                            options.parentSequence = preRevision.getSequence();
-                            options.deleted = true;
-                            options.current = preRevision.isCurrent();
-                            options.data = JSONUtils.EMPTY_JSON;
-                            options.available = false;
-                            options.copyAttachments = false;
-                            insertRevision(db,options);
-                            deletedRevision = getDocumentInQueue(db, preRevision.getId(),
-                                    newRevisionId);
-                            documentDeleted = new DocumentDeleted(preRevision, deletedRevision);
-                        }
+         if (!revisionTree.leafRevisionIds().contains(prevRevId)) {
+            throw new ConflictException("Revision to be deleted is not a leaf node:"
+                    + prevRevId);
+        }
 
-                    return deletedRevision;
+        if (!preRevision.isDeleted()) {
+            setCurrent(db, preRevision, false);
+            String newRevisionId = CouchUtils.generateNextRevisionId(preRevision.getRevision());
+            // Previous revision to be deleted could be winner revision ("current" == true),
+            // or a non-winner leaf revision ("current" == false), the new inserted
+            // revision must have the same flag as it previous revision.
+            // Deletion of non-winner leaf revision is mainly used when resolving
+            // conflicts.
+            InsertRevisionOptions options = new InsertRevisionOptions();
+            options.docNumericId = preRevision.getInternalNumericId();
+            options.revId = newRevisionId;
+            options.parentSequence = preRevision.getSequence();
+            options.deleted = true;
+            options.current = preRevision.isCurrent();
+            options.data = JSONUtils.EMPTY_JSON;
+            options.available = false;
+            options.copyAttachments = false;
+            try {
+                insertRevision(db, options);
+            } catch (AttachmentException e) {
+                //this shouldn't happen since attachments are not copied
+                throw new RuntimeException("Attachments should not be copied for a delete op",e);
```

Comment:
I'd suggest "This shouldn't happen because options.copyAttachments is set to false (as we don't do it for deletions), and AttachmentException should only occur during the copying of attachments."